sql/mdl.cc

```diff
@@ -304,6 +357,87 @@ bool Deadlock_detection_visitor::enter_node(MDL_context *node)
   return m_found_deadlock;
 }
 
+#if MARIA_PLUGIN_INTERFACE_VERSION < 0x0200
+extern "C"
+char *thd_security_context(THD *thd,
+                           char *buffer, unsigned int length,
+                           unsigned int max_query_len);
+#endif
+
+void Deadlock_detection_visitor::store_node(MDL_context *node) {
+  if (opt_print_mdl_deadlock_loop_info && !m_is_reach_max_search_depth) {
+    if (m_found_deadlock && m_n_element < MAX_SEARCH_DEPTH) {
+      const size_t MAX_THREAD_INFO_LEN = 1024;
+      const size_t MAX_LOCK_INFO_LEN = 512;
+
+      char *thread_info = new char[MAX_THREAD_INFO_LEN];
+      char *lock_info = new char[MAX_LOCK_INFO_LEN];
+
+      m_deadlock_loop_infos[m_n_element] = new MDL_deadlock_info();
+
+      size_t lock_info_buffer_size;
+      thd_security_context(node->get_owner()->get_thd(), thread_info,
+                           MAX_THREAD_INFO_LEN, 512);
+      m_deadlock_loop_infos[m_n_element]->thread_info = thread_info;
+      m_deadlock_loop_infos[m_n_element]->thread_info_len = strlen(thread_info);
+
+      node->get_wait_info(lock_info, &lock_info_buffer_size);
+      m_deadlock_loop_infos[m_n_element]->lock_info = lock_info;
+      m_deadlock_loop_infos[m_n_element]->lock_info_len = lock_info_buffer_size;
+
+      m_n_element++;
+    }
+  }
+}
+
+void Deadlock_detection_visitor::notify() {
+  char svr_error[256];
+  if (m_found_deadlock) {
+    if (m_is_reach_max_search_depth) {
+      sprintf(svr_error, "MDL deadlock occurs. %s",
+               "TOO DEEP OR LONG SEARCH IN THE LOCK TABLE WAITS-FOR GRAPH");
+      sql_print_error(svr_error);
+      return;
+    }
+    std::stringstream ss;
+    ss << "MDL deadlock occurs.Start print MDL deadlock information."; 
+    
+    char print_strs[1024];
+    size_t print_strs_size = sizeof(print_strs);
+    snprintf(print_strs, print_strs_size,
+             "\n**************MDL DeadLock Loop Info Start*************");
+    ss << (std::string)(const char*)print_strs;
```

Comment:
you do not need to cast to std::string, there is an overload for operator `<<` for the type `const  char*`
,see https://en.cppreference.com/w/cpp/io/basic_ostream/operator_ltlt2